### PR TITLE
cli/debug: debugging improvements

### PR DESCRIPTION
**Pull request by andreimatei**
See individual commits.

Release note: None

### CODE REVIEW

File: pkg/cli/debug.go [hunk 1]
```diff
@@ -449,6 +449,7 @@ func runDebugRaftLog(cmd *cobra.Command, args []string) error {
 
 	start := engine.MakeMVCCMetadataKey(keys.RaftLogPrefix(rangeID))
 	end := engine.MakeMVCCMetadataKey(keys.RaftLogPrefix(rangeID).PrefixEnd())
+	fmt.Printf("Printing keys %s -> %s (%#x - %#x)\n", start, end, start, end)
```
Comment (tbg) [inline pkg/cli/debug.go]: Doesn't do what you want -- you need to encode this MVCCMetadata key down to `[]byte` and then print it as hex.

Comment (tbg) [reply pkg/cli/debug.go]: i.e. you want the opposite of https://github.com/cockroachdb/cockroach/blob/e46eaeeb01f31bfbcbf05b623361ea2e047ddb6c/pkg/storage/engine/enginepb/decode.go#L52

Not sure if we have it on the Go side.

Comment (tbg) [reply pkg/cli/debug.go]: nvm, here it is: https://github.com/cockroachdb/cockroach/blob/50bba75d75e189f6d0496d2a940ee09c1cf82504/pkg/storage/engine/batch.go#L305

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cli/debug.go, line 452 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_P8u8RieZmFGGF4ksf-r1-452:-L_U5N1-0iEdXEmzzKdw:b54lxdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc7f5ef45ab56dce12f9d333604c51bcf6a62f/pkg/cli/debug.go#L452)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nvm, here it is: https://github.com/cockroachdb/cockroach/blob/50bba75d75e189f6d0496d2a940ee09c1cf82504/pkg/storage/engine/batch.go#L305
</blockquote></details>

done, thanks.

Btw, how did you come up with the key encoding for the `RangeAppliedState` queries you've done recently?

**Comment by andreimatei** [comment]
Pushed some more commits. PTAL.

**Review by tbg — COMMENTED** [review]
> Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/cli/debug.go, line 452 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_P8u8RieZmFGGF4ksf-r1-452:-L_qn6pv-331B2JAXCu7:bfc5qxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc7f5ef45ab56dce12f9d333604c51bcf6a62f/pkg/cli/debug.go#L452)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

done, thanks.

Btw, how did you come up with the key encoding for the `RangeAppliedState` queries you've done recently?
</blockquote></details>

In my case it was printed by the consistency checker, so I got it for free.
___
*[pkg/cli/debug/print.go, line 95 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_qnrJX0xu4clhGx5sP:-L_qnrJX0xu4clhGx5sQ:b-555knp) ([raw file](https://github.com/cockroachdb/cockroach/blob/85bf2b5ac1207ba9d4be2fd564f559d01767346e/pkg/cli/debug/print.go#L95)):*
> ```Go
> }
> 
> func decodeWriteBatch(writeBatch *storagepb.WriteBatch) (string, error) {
> ```

This is great, but I'll have to ask you to make it a little better because it's probably the last time we're touching it and we'll want it to work correctly. Applying the writebatch to an engine and then reading the engine hides stuff in the writebatch. For example, a writebatch with a deletion in it looks the same as an empty writebatch, simply because the deletion has no effect on an empty underlying engine. I think you want to use this here instead:

https://github.com/cockroachdb/cockroach/blob/50bba75d75e189f6d0496d2a940ee09c1cf82504/pkg/storage/engine/batch.go#L421-L430

The API is straightforward.
___
*[pkg/cli/debug/print.go, line 105 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_qnGjq9FhOD8CRpWGf:-L_qnGjrEHMLJv4GuIKp:bybiqjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/85bf2b5ac1207ba9d4be2fd564f559d01767346e/pkg/cli/debug/print.go#L105)):*
> ```Go
> 		return "", err
> 	}
> 	nilKey := roachpb.Key([]byte{0})
> ```

nit: not really a nil key.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cli/debug/print.go, line 95 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_qnrJX0xu4clhGx5sP:-La8QgyE-mHyaHMaVmS_:b-t6wq29) ([raw file](https://github.com/cockroachdb/cockroach/blob/85bf2b5ac1207ba9d4be2fd564f559d01767346e/pkg/cli/debug/print.go#L95)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is great, but I'll have to ask you to make it a little better because it's probably the last time we're touching it and we'll want it to work correctly. Applying the writebatch to an engine and then reading the engine hides stuff in the writebatch. For example, a writebatch with a deletion in it looks the same as an empty writebatch, simply because the deletion has no effect on an empty underlying engine. I think you want to use this here instead:

https://github.com/cockroachdb/cockroach/blob/50bba75d75e189f6d0496d2a940ee09c1cf82504/pkg/storage/engine/batch.go#L421-L430

The API is straightforward.
</blockquote></details>

done
respect to @danhhz for writing that thing

**Review by tbg — APPROVED** [review]
> Reviewed 1 of 1 files at r1, 2 of 2 files at r5, 2 of 2 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/cli/debug/print.go, line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-LaGLIbU1Kyo8wKqsyhG:-LaGLIbV1-AxNI0cJDRM:b-lxpu85) ([raw file](https://github.com/cockroachdb/cockroach/blob/4245a8b0ebace79972a7c1a4453bd0668a6e8b9f/pkg/cli/debug/print.go#L42)):*
> ```Go
> }
> 
> // SprintKey prety-prings the specified MVCCKey.
> ```

pretty
___
*[pkg/cli/debug/print.go, line 113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-LaGLOPO78dX_sWpsvvz:-LaGLOPO78dX_sWpsvw-:bh31dc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4245a8b0ebace79972a7c1a4453bd0668a6e8b9f/pkg/cli/debug/print.go#L113)):*
> ```Go
> 		switch r.BatchType() {
> 		case engine.BatchTypeDeletion:
> 			mvccKey, err := r.MVCCKey()
> ```

Isn't it silly that you're doing this in every block?

**Review by andreimatei — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-:-LaGfqDl9GCOpyj8qtgN:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cli/debug/print.go, line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-LaGLIbU1Kyo8wKqsyhG:-LaGeCDX3lFOLxumjK_c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4245a8b0ebace79972a7c1a4453bd0668a6e8b9f/pkg/cli/debug/print.go#L42)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

pretty
</blockquote></details>

Done.
___
*[pkg/cli/debug/print.go, line 113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-LaGLOPO78dX_sWpsvvz:-LaGeghS60Cm6Sl5LBWg:b-s7w3c9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4245a8b0ebace79972a7c1a4453bd0668a6e8b9f/pkg/cli/debug/print.go#L113)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Isn't it silly that you're doing this in every block?
</blockquote></details>

it seems silly but I'm not sure how best to deal with the `default` case below. Anyway, whatevs, I'll leave it.

**Comment by tbg** [comment]
@andreimatei could you backport? Will be good to have this around.

**Comment by andreimatei** [comment]
backport is #35901

**Pull request merged**